writer/firehose_writer.go

```diff
@@ -59,26 +85,116 @@ func NewFirehoseWriter(config FirehoseWriterConfig) (*FirehoseWriter, error) {
-func (f *FirehoseWriter) ProcessMessage(msg string) error {
-	atomic.AddInt64(&f.recvRecordCount, 1)
+func (f *FirehoseWriter) Initialize(shardID string) error {
+	f.shardID = shardID
+	f.lastCheckpoint = time.Now()
+	return nil
+}
+
+func (f *FirehoseWriter) checkpoint(checkpointer kcl.Checkpointer, sequenceNumber string, subSequenceNumber int) {
+	for n := -1; n < f.checkpointRetries; n++ {
+		err := checkpointer.Checkpoint(sequenceNumber, subSequenceNumber)
+		if err == nil {
+			return
+		}
+
+		if cperr, ok := err.(kcl.CheckpointError); ok {
+			switch cperr.Error() {
+			case "ShutdownException":
+				fmt.Fprintf(os.Stderr, "Encountered shutdown exception, skipping checkpoint\n")
+				return
+			case "ThrottlingException":
+				fmt.Fprintf(os.Stderr, "Was throttled while checkpointing, will attempt again in %s", f.sleepDuration)
+			case "InvalidStateException":
+				fmt.Fprintf(os.Stderr, "MultiLangDaemon reported an invalid state while checkpointing\n")
+			default:
+				fmt.Fprintf(os.Stderr, "Encountered an error while checkpointing: %s", err)
+			}
+		}
+
+		if n == f.checkpointRetries {
+			fmt.Fprintf(os.Stderr, "Failed to checkpoint after %d attempts, giving up.\n", f.checkpointRetries)
+			return
+		}
+
+		time.Sleep(f.sleepDuration)
+	}
+}
+
+func (f *FirehoseWriter) ProcessRecords(records []kcl.Record, checkpointer kcl.Checkpointer) error {
+	for _, record := range records {
+		// Wait until rate limiter permits one more record to be processed
+		f.rateLimiter.Wait(context.Background())
+		atomic.AddInt64(&f.recvRecordCount, 1)
+		err := f.processRecord(record)
+		if err != nil {
+			atomic.AddInt64(&f.failedRecordCount, 1)
+			continue
+		}
+	}
+
+	// Checkpoint Kinesis stream
+	if time.Now().Sub(f.lastCheckpoint) > f.checkpointFreq {
+		f.checkpoint(checkpointer, f.largestSeqFlushed.String(), f.largestSubSeqFlushed)
+		f.lastCheckpoint = time.Now()
+
+		// Write status to file
+		appendToFile(f.logFile, fmt.Sprintf("%s -- Received:%d Sent:%d Failed:%d\n", f.shardID, f.recvRecordCount, f.sentRecordCount, f.failedRecordCount))
+	}
+
+	return nil
+}
+
+func (f *FirehoseWriter) processRecord(record kcl.Record) error {
+	// Base64 decode the record
+	data, err := base64.StdEncoding.DecodeString(record.Data)
+	if err != nil {
+		return err
+	}
 
 	// TODO: Fully decode the message
 	fields := map[string]interface{}{
-		"rawlog": msg,
+		"rawlog": string(data),
+	}
+	msg, err := json.Marshal(fields)
+	if err != nil {
+		return err
 	}
 
-	record, err := json.Marshal(fields)
+	err = f.messageBatcher.AddMessage(msg, record.SequenceNumber, record.SubSequenceNumber)
 	if err != nil {
-		atomic.AddInt64(&f.failedRecordCount, 1)
 		return err
 	}
 
-	return f.messageBatcher.Send(record)
+	return nil
 }
 
-// Flush writes a batch of records to AWS Firehose
-func (f *FirehoseWriter) Flush(batch [][]byte) {
+func appendToFile(filename, text string) error {
```

Comment:
nit but this seems potentially very inefficient if there's a high volume of logs--could the config take in a `io.Writer` instead?

Follow-up nit is that if there are multiple FirehoseWriters sharing a single file, writes need to be coordinated, in which case taking in a `log.Logger` might make the most sense (`log.Logger` is threadsafe)